### ISSUE TITLE
Errors from ServiceControl.Monitoring instance v5.5 ingesting NServiceBus metrics

### ISSUE THREAD

**Issue by saschanm** [Particular/ServiceControl#4333]
#### Description

After upgrading Service Control instances from v5.4.1 to v5.5.0 errors started appearing in ServicePulse from the Monitoring instance and logs started filling with errors.

Behaviour only seems to be apparent for metrics coming from old NSB v5 endpoints.
NSB v7 endpoints seem unaffected.

#### Expected behavior
Metrics should be consumed by Monitoring instance without error

#### Actual behavior
Logs filling with repeated error messages until we downgraded Monitoring instance to v5.4.1

#### Versions
ServiceControl.Monitoring Instance (5.5.0)

Metrics error from endpoints using:
    NServiceBus (5.2.26)
    NServiceBus.Metrics.ServiceControl (1.3.1)
    NServiceBus.Metrics.ServiceControl.Msmq (1.0.1)
    
Metrics seem ok from endpoints using:
   NServiceBus (7.7)
   NServiceBus.Metrics (3.1)
   NServiceBus.Metrics.ServiceControl (3.0.6)
   NServiceBus.Metrics.ServiceControl.Msmq (3.0.1)

### Steps to reproduce

Install Service Control v5.5.0 including a monitoring instance

Configure a new Nsb endpoint with metrics using:
    NServiceBus (5.2.26)
    NServiceBus.Metrics.ServiceControl (1.3.1)
    NServiceBus.Metrics.ServiceControl.Msmq (1.0.1)
    
configured the endpoint to send metrics to the queues configured for the monitoring endpoint.
configure a message and handler for the new endpoint.
schedule the command to send every few seconds to ensure CriticalTime and ProcessingTime metrics are sent to the monitoring queue.
Those messages will fail with errors as shown in the relevant log output below

 

### Relevant log output

```shell
2024-07-24 16:50:11.2597|79|Error|NServiceBus.MoveToError|Moving message '64a36db4-96ce-477c-8365-afead58636b2' to the error queue 'error' because processing failed due to an exception:|System.ArgumentException: An item with the same key has already been added. Key: nservicebus.message_handler_types
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at NServiceBus.LoadHandlersConnector.Invoke(IIncomingLogicalMessageContext context, Func`2 stage) in /_/src/NServiceBus.Core/Pipeline/Incoming/LoadHandlersConnector.cs:line 41
   at ServiceControl.Monitoring.MessagePoolReleasingBehavior.Invoke(IIncomingLogicalMessageContext context, Func`1 next) in /_/src/ServiceControl.Monitoring/MessagePoolReleasingBehavior.cs:line 14
   at NServiceBus.DeserializeMessageConnector.Invoke(IIncomingPhysicalMessageContext context, Func`2 stage) in /_/src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs:line 39
   at NServiceBus.SubscriptionReceiverBehavior.Invoke(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/SubscriptionReceiverBehavior.cs:line 29
   at NServiceBus.ProcessingStatisticsBehavior.Invoke(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/Performance/Statistics/ProcessingStatisticsBehavior.cs:line 25
   at NServiceBus.TransportReceiveToPhysicalMessageConnector.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageConnector.cs:line 36
   at NServiceBus.RetryAcknowledgementBehavior.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/ServicePlatform/Retries/RetryAcknowledgementBehavior.cs:line 25
   at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext, CancellationToken cancellationToken) in /_/src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:line 50
   at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext, CancellationToken cancellationToken) in /_/src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:line 78
   at NServiceBus.Transport.Msmq.ReceiveStrategy.TryProcessMessage(String messageId, Dictionary`2 headers, ReadOnlyMemory`1 body, TransportTransaction transaction, ContextBag context, CancellationToken cancellationToken) in /_1/nservicebus.transport.msmq.sources/3.0.1/contentFiles/cs/net8.0-windows7.0/NSB.Transport.Msmq/ReceiveStrategy.cs:line 112
   at NServiceBus.Transport.Msmq.ReceiveOnlyNativeTransactionStrategy.ProcessMessage(Message message, Dictionary`2 headers, ContextBag context, CancellationToken cancellationToken) in /_1/nservicebus.transport.msmq.sources/3.0.1/contentFiles/cs/net8.0-windows7.0/NSB.Transport.Msmq/ReceiveOnlyNativeTransactionStrategy.cs:line 92
2024-07-24 16:50:11.2597|69|Error|NServiceBus.MoveToError|Moving message '0aff5105-802c-453c-ba97-e5a1647ba5fe' to the error queue 'error' because processing failed due to an exception:|System.ArgumentException: An item with the same key has already been added. Key: nservicebus.message_handler_types
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at NServiceBus.LoadHandlersConnector.Invoke(IIncomingLogicalMessageContext context, Func`2 stage) in /_/src/NServiceBus.Core/Pipeline/Incoming/LoadHandlersConnector.cs:line 41
   at ServiceControl.Monitoring.MessagePoolReleasingBehavior.Invoke(IIncomingLogicalMessageContext context, Func`1 next) in /_/src/ServiceControl.Monitoring/MessagePoolReleasingBehavior.cs:line 14
   at NServiceBus.DeserializeMessageConnector.Invoke(IIncomingPhysicalMessageContext context, Func`2 stage) in /_/src/NServiceBus.Core/Pipeline/Incoming/DeserializeMessageConnector.cs:line 39
   at NServiceBus.SubscriptionReceiverBehavior.Invoke(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/SubscriptionReceiverBehavior.cs:line 29
   at NServiceBus.ProcessingStatisticsBehavior.Invoke(IIncomingPhysicalMessageContext context, Func`2 next) in /_/src/NServiceBus.Core/Performance/Statistics/ProcessingStatisticsBehavior.cs:line 25
   at NServiceBus.TransportReceiveToPhysicalMessageConnector.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/Pipeline/Incoming/TransportReceiveToPhysicalMessageConnector.cs:line 36
   at NServiceBus.RetryAcknowledgementBehavior.Invoke(ITransportReceiveContext context, Func`2 next) in /_/src/NServiceBus.Core/ServicePlatform/Retries/RetryAcknowledgementBehavior.cs:line 25
   at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext, CancellationToken cancellationToken) in /_/src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:line 50
   at NServiceBus.MainPipelineExecutor.Invoke(MessageContext messageContext, CancellationToken cancellationToken) in /_/src/NServiceBus.Core/Pipeline/MainPipelineExecutor.cs:line 78
   at NServiceBus.Transport.Msmq.ReceiveStrategy.TryProcessMessage(String messageId, Dictionary`2 headers, ReadOnlyMemory`1 body, TransportTransaction transaction, ContextBag context, CancellationToken cancellationToken) in /_1/nservicebus.transport.msmq.sources/3.0.1/contentFiles/cs/net8.0-windows7.0/NSB.Transport.Msmq/ReceiveStrategy.cs:line 112
   at NServiceBus.Transport.Msmq.ReceiveOnlyNativeTransactionStrategy.ProcessMessage(Message message, Dictionary`2 headers, ContextBag context, CancellationToken cancellationToken) in /_1/nservicebus.transport.msmq.sources/3.0.1/contentFiles/cs/net8.0-windows7.0/NSB.Transport.Msmq/ReceiveOnlyNativeTransactionStrategy.cs:line 92
```

#### Workarounds

Remove the Monitoring instance and reinstall a version 5.4.1 instance. ServiceControl and Audit instances can remain v5.5.0 AFAICT

**Comment by NChaganlal** [comment]
Hi @saschanm,

Would you mind opening a formal case on non-critical support through the following [URL](https://particular.net/support)? Please choose the "Request non-critical support option". Thank you.

**Comment by gursharan001** [comment]
Hi there, we see the exact same issue with NSB7 endpoints on Azure ServiceBus transport as well

**Comment by mauroservienti** [comment]
We have identified the root cause of the issue, and we are working on a fix.

**Comment by jpalac** [comment]
Fix deployed in [ServiceControl 5.5.2](https://github.com/Particular/ServiceControl/releases/tag/5.5.2)
